src/storage-client/src/controller.rs

```diff
@@ -2295,6 +2297,25 @@ where
                     frontier.clone(),
                 )]));
             }
+
+            // Check if the collection is for a Webhook source, unregister if so.
+            let collection = self.state.collections.get(&id);
+            if let Some(CollectionState { description, .. }) = collection {
+                if description.data_source == DataSource::Webhook && frontier.is_empty() {
+                    // Unregister our collection from the manager so writes should no longer occur.
+                    self.state.collection_manager.unregsiter_collection(id);
+                    // Wait for a barrier to assert that no work is in progress for the collection.
+                    self.state.collection_manager.barrier().await;
```

Comment:
I'm not _totally sure_ if waiting for a barrier here is necessary, i.e. if there are in-progress writes for a shard it would be a problem to concurrently try to finalize/drop it. But I'm more comfortable having it here because for sure after waiting for a barrier the `CollectionManager` would always reject writes for a collection, since we unregistered.

Reply:
Appreciated the context you added about the barrier.

Re-reading this code, it looks like the barrier is actually necessary because we're changing (minimizing) the time for which the lock over the collections are held, e.g. a shard could be dropped while an update is occurring. That would have been bad and your foresight to handle this is 👌 . 

It looks like the major benefit of shortening the lifetime of the lock being held in the `CollectionManager` is that it will expedite adding or dropping webhook sources--is that your thinking here, or were there other more overarching concerns?

I have some apprehension that this might be a premature optimization that introduces a fair bit of complexity (two synchronization primitives instead of one, one of which is fairly bespoke); we haven't yet heard that there are any issues with the length of time that the lock is held so we're paying for the complexity of a solution that we might never otherwise need (and once it's in, it's likely no one is ever going to question its necessity again).

How does this square with your thinking? lmk if you'd like a higher-bandwidth chat.

Reply:
For posterity, Sean and I huddled and realized we shouldn't hold the lock for `collections` across the `.await` point of monotonic append, so we need some other synchronization primitive _or_ have persist write handles gracefully handle missing shards. We tried the later but found the API of a single response for the entire batch to be limiting. Specifically, it was difficult to return an error for just a single id that might not exist, we left a TODO to revist this.

Our final solution was to move the barrier logic (note: renamed to notify since it wasn't exactly a barrier) into `unregister_collection(...)`, instead of having callers manage it themselves.